db/db_impl_compaction_flush.cc

```diff
@@ -1199,22 +1302,43 @@ void DBImpl::MaybeScheduleFlushOrCompaction() {
   auto bg_job_limits = GetBGJobLimits();
   bool is_flush_pool_empty =
       env_->GetBackgroundThreads(Env::Priority::HIGH) == 0;
-  while (!is_flush_pool_empty && unscheduled_flushes_ > 0 &&
-         bg_flush_scheduled_ < bg_job_limits.max_flushes) {
-    unscheduled_flushes_--;
-    bg_flush_scheduled_++;
-    env_->Schedule(&DBImpl::BGWorkFlush, this, Env::Priority::HIGH, this);
-  }
-
-  // special case -- if high-pri (flush) thread pool is empty, then schedule
-  // flushes in low-pri (compaction) thread pool.
-  if (is_flush_pool_empty) {
-    while (unscheduled_flushes_ > 0 &&
-           bg_flush_scheduled_ + bg_compaction_scheduled_ <
-               bg_job_limits.max_flushes) {
+  if (atomic_flush_) {
+    if (!is_flush_pool_empty && unscheduled_flushes_ > 0 &&
+        bg_flush_scheduled_ < bg_job_limits.max_flushes) {
+      unscheduled_flushes_ = 0;
+      bg_flush_scheduled_++;
+      env_->Schedule(&DBImpl::BGWorkFlush, this, Env::Priority::HIGH, this);
+    }
+
+    // special case -- if high-pri (flush) thread pool is empty, then schedule
+    // flushes in low-pri (compaction) thread pool.
+    if (is_flush_pool_empty) {
+      if (unscheduled_flushes_ > 0 &&
+          bg_flush_scheduled_ + bg_compaction_scheduled_ <
+          bg_job_limits.max_flushes) {
+        unscheduled_flushes_ = 0;
```

Comment:
can there be multiple unscheduled group flushes?

Reply:
There can be. That's why we remember the memtable id for each column family, indicating the highest memtable id of this column family during this group flush req.

Reply:
Makes sense. Is it possible for `unscheduled_flushes_` to be greater than one here, though? If so I am not sure about resetting it to zero upon scheduling the first one.

Reply:
You are right.
Originally I also used `unscheduled_flushes_` to track a group of flushes that must happen atomically. These flushes also reside in the normal flush queue. Once a flush is initiated, all of them must be scheduled, that's why I reset the counter to 0.
Now we have a separate queue for group flush reqs, let me investigate whether it still makes sense to share the same `unscheduled_flushes_` variable.

Reply:
Fixed, will push the change later.